_posts/2020-12-27-spring-boot-exception-hanlding.md

```diff
@@ -0,0 +1,596 @@
+corresponding error message.
+
+To address this we can we annotate our Exception class with `@ResponseStatus` and pass in the desired HTTP response status
+in its value property:
+
+```java
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class NoSuchElementFound extends RuntimeException {
+
+}
+```
+
+This change will result in a much better response:
+
+```json
+{
+    "timestamp": "2020-11-29T09:42:12.287+00:00",
+    "status": 404,
+    "error": "Not Found",
+    "message": "Item with id 1 not found",
+    "path": "/product/1"
+} 
+```
+
+Another way to achieve the same is by extending `ResponseStatusException` class:
+
+```java
+public class NoSuchElementFoundException extends ResponseStatusException {
+
+    public NoSuchElementFoundException(String message){
+        super(HttpStatus.NOT_FOUND, message);
+    }
+}
+```
+
+This approach comes in handy when we want to manipulate the response headers too.
+
+Now, as we can see `@ResponseStatus` along with `server.error` properties allows us to manipulate almost all the fields
+in our Spring defined error response payload but, what if want to manipulate response payload structure too? Let's see how
+we can achieve that in the next section. 
+
+## `@ExceptionHandler`
+`@ExceptionHandler` annotation gives us a lot of flexibility in terms of handling exceptions. For starters, to use it, we
+simply need to create a method either in the controller itself or in a `@ControllerAdvice` class and
+annotate it with `@ExceptionHandler`.
+
+```java
+@RestController
+@RequestMapping("/product")
+public class ProductController {
+    private final ProductService productService;
+    //constructor omitted for brevity...
+
+    @GetMapping("/{id}")
+    public Response getProduct(@PathVariable String id) {
+        return productService.getProduct(id);
+    }
+
+    @ExceptionHandler(NoSuchElementFoundException.class)
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public ResponseEntity<String> handleNoSuchElementFoundException(
+            NoSuchElementFoundException exception
+    ) {
+        return ResponseEntity
+                .status(httpStatus)
+                .body(exception.getMessage());
+    }
+
+}
+```
+
+`@ExceptionHandler` takes in an exception, or a list of exceptions as an argument that we want to handle in the defined
+method. If we don't wish to do that then simply defining exception as a parameter of the method will also do:
+
+```java
+@ExceptionHandler
+public ResponseEntity<String> handleNoSuchElementFoundException(
+        NoSuchElementFoundException exception
+        )
+```
+
+Or we can keep both just as shown in the example controller above. It gives better readability. For the same reason, we have
+kept `@ResponseStatus(HttpStatus.NOT_FOUND)` on the handler method. Although it's not required as we have already provided HTTP status in `ResponseEntity`.
```

Comment:
This confuses me. I'm guessing that if we use `@ResponseStatus`, we can "inject" the `httpStatus` variable as a parameter into the method? The code snippet and explanation doesn't make sense as it currently stands.